lib/pos_blockchain.go

```diff
@@ -0,0 +1,1331 @@
+	for _, ancestor := range lineageFromCommittedTip {
+		bc.addBlockToBestChain(ancestor)
+	}
+	// Add the new tip to the best chain.
+	bc.addBlockToBestChain(blockNode)
+	return true, nil
+}
+
+// shouldReorg determines if we should reorg to the block provided. We should reorg if
+// this block is proposed in a view greater than or equal to the currentView. Other
+// functions have validated that this block is not extending from a committed block
+// that is not the latest committed block, so there is no need to validate that here.
+func (bc *Blockchain) shouldReorg(blockNode *BlockNode, currentView uint64) bool {
+	chainTip := bc.BlockTip()
+	// If this block extends from the chain tip, there's no need to reorg.
+	if chainTip.Hash.IsEqual(blockNode.Header.PrevBlockHash) {
+		return false
+	}
+	// If the block is proposed in a view less than the current view, there's no need to reorg.
+	return blockNode.Header.ProposedInView >= currentView
+}
+
+// addBlockToBestChain adds the block to the best chain.
+func (bc *Blockchain) addBlockToBestChain(blockNode *BlockNode) {
+	bc.bestChain = append(bc.bestChain, blockNode)
+	bc.bestChainMap[*blockNode.Hash] = blockNode
+}
+
+// runCommitRuleOnBestChain commits the grandparent of the block if possible.
+// Specifically, this updates the CommittedBlockStatus of its grandparent
+// and flushes the view after connecting the grandparent block to the DB.
+func (bc *Blockchain) runCommitRuleOnBestChain() error {
+	currentBlock := bc.BlockTip()
+	// If we can commit the grandparent, commit it.
+	// Otherwise, we can't commit it and return nil.
+	blockToCommit, canCommit := bc.canCommitGrandparent(currentBlock)
+	if !canCommit {
+		return nil
+	}
+	// Find all uncommitted ancestors of block to commit
+	_, idx := bc.getCommittedTip()
+	if idx == -1 {
+		// This is an edge case we'll never hit in practice since all the PoW blocks
+		// are committed.
+		return errors.New("runCommitRuleOnBestChain: No committed blocks found")
+	}
+	uncommittedAncestors := []*BlockNode{}
+	for ii := idx + 1; ii < len(bc.bestChain); ii++ {
+		uncommittedAncestors = append(uncommittedAncestors, bc.bestChain[ii])
+		if bc.bestChain[ii].Hash.IsEqual(blockToCommit) {
+			break
+		}
+	}
+	for ii := 0; ii < len(uncommittedAncestors); ii++ {
+		if err := bc.commitBlockPoS(uncommittedAncestors[ii].Hash); err != nil {
+			return errors.Wrapf(err, "runCommitRuleOnBestChain: Problem committing block %v", uncommittedAncestors[ii].Hash.String())
+		}
+	}
+	return nil
+}
+
+// canCommitGrandparent determines if the grandparent of the current block can be committed.
+// The grandparent can be committed if there exists a direct parent-child relationship
+// between the grandparent and parent of the new block, meaning the grandparent and parent
+// are proposed in consecutive views, and the "parent" is an ancestor of the incoming block (not necessarily consecutive views).
+// Additionally, the grandparent must not already be committed.
```

Comment:
Would be nice to add a reference to Revolution PoS doc here so people understand where this rule comes from.

Reply:
Also this is definitely the commit rule I remember. But something I was thinking about is what happens if you have a chain like the following, where the x'es mean "timeout occurred" for that view:
```
A -> x -> x -> B -> x -> x -> C -> x -> x -> D -> x -> x -> E
```

I think you could have an infinitely long chain of this where blocks would just never commit. Is that true? If so, this feels wrong. It feels like we should be able to commit if there are three blocks of QC built up, even if there are timeouts between them. I would ask Piotr about it, as he went the deepest on this but I can also dig into the stuff if you can't find the answer.

The reason why this is important btw is we want to be able to tell exchanges that they can consider a certain number of blocks as committed. And it's kindof annoying if, in theory, it could take an infinite number of blocks to commit.